In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("saurabhshahane/arabic-classification")

print("Path to dataset files:", path)

100%|██████████| 83.4M/83.4M [00:01<00:00, 67.6MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/saurabhshahane/arabic-classification/versions/1


In [ ]:
import nltk
import pandas as pd
import numpy as np

print("Libraries imported: nltk, pandas, numpy")

Libraries imported: nltk, pandas, numpy


In [ ]:
import os

# List the contents of the downloaded dataset path
print(os.listdir(path))

['arabic_dataset_classifiction.csv']


In [ ]:
import os

# Construct the initial path to what we expect to be the CSV file
csv_file_path = os.path.join(path, 'arabic_dataset_classifiction.csv')

# Check if the constructed path is actually a directory
if os.path.isdir(csv_file_path):
    # If it's a directory, list its contents to find the actual CSV file
    files_in_sub_dir = os.listdir(csv_file_path)
    print(f"Found directory at '{csv_file_path}'. Contents: {files_in_sub_dir}")

    # Assuming the actual CSV file has the same name and is directly inside this directory
    # Or, if there's only one file, assume that's it.
    actual_csv_filename = None
    for f_name in files_in_sub_dir:
        if f_name.endswith('.csv'):
            actual_csv_filename = f_name
            break

    if actual_csv_filename:
        csv_file_path = os.path.join(csv_file_path, actual_csv_filename)
        print(f"Attempting to read from: {csv_file_path}")
    else:
        raise FileNotFoundError(f"No CSV file found inside the directory: {csv_file_path}")

# Read the CSV file into a DataFrame
df = pd.read_csv(csv_file_path)

print("DataFrame Head:")
print(df.head())

print("\nDataFrame Statistics:")
print(df.describe())

Found directory at '/root/.cache/kagglehub/datasets/saurabhshahane/arabic-classification/versions/1/arabic_dataset_classifiction.csv'. Contents: ['arabic_dataset_classifiction.csv']
Attempting to read from: /root/.cache/kagglehub/datasets/saurabhshahane/arabic-classification/versions/1/arabic_dataset_classifiction.csv/arabic_dataset_classifiction.csv
DataFrame Head:
                                                text  targe
0  بين أستوديوهات ورزازات وصحراء مرزوكة وآثار ولي...      0
1  قررت النجمة الأمريكية أوبرا وينفري ألا يقتصر ع...      0
2  أخبارنا المغربية الوزاني تصوير الشملالي ألهب ا...      0
3  اخبارنا المغربية قال ابراهيم الراشدي محامي سعد...      0
4  تزال صناعة الجلود في المغرب تتبع الطريقة التقل...      0

DataFrame Statistics:
               targe
count  111728.000000
mean        2.620659
std         1.451010
min         0.000000
25%         1.000000
50%         3.000000
75%         4.000000
max         4.000000


In [ ]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import re

# Download necessary NLTK data if not already present
# 'stopwords' for stop word removal
# 'punkt' and 'punkt_tab' for tokenization
nltk.download('stopwords', quiet=True)
nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True) # Explicitly download punkt_tab for Arabic tokenization

print("NLTK stopwords, punkt, and punkt_tab data are available.")

NLTK stopwords, punkt, and punkt_tab data are available.


In [ ]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import re

# Download necessary NLTK data if not already present
# 'stopwords' for stop word removal
# 'punkt' and 'punkt_tab' for tokenization
nltk.download('stopwords', quiet=True)
nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True) # Explicitly download punkt_tab for Arabic tokenization

print("NLTK stopwords, punkt, and punkt_tab data are available.")

NLTK stopwords, punkt, and punkt_tab data are available.


In [ ]:
import collections
import re
import nltk
from nltk.corpus import stopwords

# Download necessary NLTK data
nltk.download('stopwords', quiet=True)
nltk.download('punkt', quiet=True)

# Define Arabic stop words
arabic_stopwords = set(stopwords.words('arabic'))

def preprocess(text):
    # 1. Remove Tatweel and diacritics (harakat)
    text = re.sub(r'\u064b|\u064c|\u064d|\u064e|\u064f|\u0650|\u0651|\u0652|\u0653|\u0670', '', text) # Remove diacritics
    text = re.sub(r'\u0640', '', text) # Remove Tatweel

    # 2. Normalize some character forms
    text = re.sub(r'[\u0622\u0623\u0625]', '\u0627', text) # Normalize Alif forms to simple Alif
    text = re.sub(r'\u0649', '\u064a', text) # Normalize Alef Maksura to Yeh
    text = re.sub(r'\u0629', '\u0647', text) # Normalize Ta Marbuta to Ha

    # 3. Remove non-Arabic characters, numbers, and punctuation
    text = re.sub(r'[^\u0621-\u064A\s]', '', text) # Keep only Arabic letters and spaces
    text = re.sub(r'\s+', ' ', text).strip() # Replace multiple spaces with a single space
    return text

def tokenize(text):
    # Split the input text into a list of tokens by whitespace
    tokens = text.split()
    # Filter out tokens that are present in the arabic_stopwords set
    tokens = [word for word in tokens if word not in arabic_stopwords]
    return tokens

def preprocess_input(sentence, word_to_id, max_len):
    # Apply preprocessing and tokenization
    cleaned_text = preprocess(sentence)
    tokens = tokenize(cleaned_text)

    # Convert tokens to numerical IDs, handle unknown words with 0
    indexed_tokens = [word_to_id.get(token, 0) for token in tokens]

    # Pad or truncate the sequence to max_len
    if len(indexed_tokens) > max_len:
        indexed_tokens = indexed_tokens[:max_len]
    else:
        indexed_tokens = indexed_tokens + [0] * (max_len - len(indexed_tokens))
    return indexed_tokens

def concatenate(list_of_lists):
    # Flatten a list of lists into a single list
    return [item for sublist in list_of_lists for item in sublist]

def data_vocab(text_series):
    all_words = []
    for text in text_series:
        # Assuming each text is already preprocessed and is a string of space-separated tokens
        all_words.extend(text.split())
    return collections.Counter(all_words)

print("Preprocessing functions and utilities defined.")


Preprocessing functions and utilities defined.


**Reasoning**:
Now that the preprocessing functions are defined, I need to apply the `preprocess` function to the 'text' column of the DataFrame `df` to create a new column named 'cleaned_text'. This is the next logical step before generating the word frequency vocabulary.



**Reasoning**:
The previous error `TypeError: expected string or bytes-like object, got 'float'` indicates that the 'text' column contains non-string values (likely NaNs). To resolve this, I will fill any NaN values in the 'text' column with an empty string before applying the `preprocess` function.



In [ ]:
df['cleaned_text'] = df['text'].fillna('').apply(preprocess)

print("DataFrame with 'cleaned_text' column:")
print(df.head())

DataFrame with 'cleaned_text' column:
                                                text  targe  \
0  بين أستوديوهات ورزازات وصحراء مرزوكة وآثار ولي...      0   
1  قررت النجمة الأمريكية أوبرا وينفري ألا يقتصر ع...      0   
2  أخبارنا المغربية الوزاني تصوير الشملالي ألهب ا...      0   
3  اخبارنا المغربية قال ابراهيم الراشدي محامي سعد...      0   
4  تزال صناعة الجلود في المغرب تتبع الطريقة التقل...      0   

                                        cleaned_text  
0  بين استوديوهات ورزازات وصحراء مرزوكه واثار ولي...  
1  قررت النجمه الامريكيه اوبرا وينفري الا يقتصر ع...  
2  اخبارنا المغربيه الوزاني تصوير الشملالي الهب ا...  
3  اخبارنا المغربيه قال ابراهيم الراشدي محامي سعد...  
4  تزال صناعه الجلود في المغرب تتبع الطريقه التقل...  


**Reasoning**:
Now that the 'cleaned_text' column is available, the next step is to tokenize these cleaned texts and then generate a vocabulary using the `data_vocab` function. This will prepare the data for numerical representation.



In [ ]:
df['tokenized_text'] = df['cleaned_text'].apply(tokenize)

# Generate vocabulary
vocabulary = data_vocab(df['tokenized_text'].apply(lambda x: ' '.join(x)))

print("DataFrame with 'tokenized_text' column:")
print(df.head())

print("\nTop 10 most common words in vocabulary:")
print(vocabulary.most_common(10))

DataFrame with 'tokenized_text' column:
                                                text  targe  \
0  بين أستوديوهات ورزازات وصحراء مرزوكة وآثار ولي...      0   
1  قررت النجمة الأمريكية أوبرا وينفري ألا يقتصر ع...      0   
2  أخبارنا المغربية الوزاني تصوير الشملالي ألهب ا...      0   
3  اخبارنا المغربية قال ابراهيم الراشدي محامي سعد...      0   
4  تزال صناعة الجلود في المغرب تتبع الطريقة التقل...      0   

                                        cleaned_text  \
0  بين استوديوهات ورزازات وصحراء مرزوكه واثار ولي...   
1  قررت النجمه الامريكيه اوبرا وينفري الا يقتصر ع...   
2  اخبارنا المغربيه الوزاني تصوير الشملالي الهب ا...   
3  اخبارنا المغربيه قال ابراهيم الراشدي محامي سعد...   
4  تزال صناعه الجلود في المغرب تتبع الطريقه التقل...   

                                      tokenized_text  
0  [استوديوهات, ورزازات, وصحراء, مرزوكه, واثار, و...  
1  [قررت, النجمه, الامريكيه, اوبرا, وينفري, الا, ...  
2  [اخبارنا, المغربيه, الوزاني, تصوير, الشملالي, ...  
3  [اخبارنا, المغربيه, ق

**Reasoning**:
To convert the tokenized words into numerical representations, I will create two dictionaries: one for mapping words to unique integer IDs and another for mapping IDs back to words. I'll also define the vocabulary size and a maximum sequence length, which are crucial for preparing the data for model input.



In [ ]:
word_to_id = {word: i + 1 for i, (word, _) in enumerate(vocabulary.most_common())}
id_to_word = {i + 1: word for i, (word, _) in enumerate(vocabulary.most_common())}

vocab_size = len(word_to_id) + 1 # +1 for padding/unknown words
max_len = 100 # Define a maximum sequence length for input to the model

print(f"Vocabulary size: {vocab_size}")
print(f"Maximum sequence length: {max_len}")
print("First 10 word to ID mappings:")
for word, id_val in list(word_to_id.items())[:10]:
    print(f"{word}: {id_val}")


Vocabulary size: 388617
Maximum sequence length: 100
First 10 word to ID mappings:
ان: 1
علي: 2
الي: 3
خلال: 4
انه: 5
المغرب: 6
الفريق: 7
المغربيه: 8
الوطني: 9
اجل: 10


**Reasoning**:
With the `word_to_id` mapping, `vocab_size`, and `max_len` defined, I will now apply the `preprocess_input` function to the 'cleaned_text' column of the DataFrame to create numerical representations of the text, ready for model training. This function will convert the tokenized words into their corresponding integer IDs and ensure all sequences have a uniform length.



In [ ]:
df['indexed_text'] = df['cleaned_text'].apply(lambda x: preprocess_input(x, word_to_id, max_len))

print("DataFrame with 'indexed_text' column (first 5 rows and first 10 elements of the list):")
for i in range(5):
    print(f"Row {i}: {df['indexed_text'].iloc[i][:10]}")

DataFrame with 'indexed_text' column (first 5 rows and first 10 elements of the list):
Row 0: [18319, 5236, 88993, 42856, 6229, 15203, 278, 12244, 2355, 1417]
Row 1: [723, 6193, 558, 33825, 49564, 57, 6732, 4024, 2, 1182]
Row 2: [111, 8, 4914, 1696, 43506, 37844, 1450, 18, 9655, 9711]
Row 3: [111, 8, 54, 1548, 19486, 6184, 808, 5800, 1, 7342]
Row 4: [3035, 1867, 62693, 6, 3655, 1738, 1633, 181400, 58252, 3076]


## Demonstrate Masking and Concatenation

### Subtask:
Demonstrate the usage of `preprocess_input` and `concatenate` functions with a sample sentence and the generated vocabulary.


**Reasoning**:
To demonstrate the `preprocess_input` and `concatenate` functions as requested, I will define a sample Arabic sentence and a sample list of lists, then apply these functions and print their outputs.



In [ ]:
import numpy as np

# 1. Define a sample Arabic sentence
sample_sentence = "هذا مثال على جملة عربية لتجربة دالة المعالجة المسبقة."

# 2. Apply the preprocess_input function
indexed_sample_sentence = preprocess_input(sample_sentence, word_to_id, max_len)

# 3. Print the original sample sentence and its indexed_text representation
print(f"Original sample sentence: {sample_sentence}")
print(f"Indexed sample sentence (first 10 elements): {indexed_sample_sentence[:10]}")
print(f"Indexed sample sentence (full length): {len(indexed_sample_sentence)}")

# 4. Create a list of lists
list_of_lists_example = [[1, 2, 3], [4, 5, 6], [7]]

# 5. Apply the concatenate function
flattened_list = concatenate(list_of_lists_example)

# 6. Print the original list of lists and the flattened list
print(f"\nOriginal list of lists: {list_of_lists_example}")
print(f"Flattened list using concatenate: {flattened_list}")

Original sample sentence: هذا مثال على جملة عربية لتجربة دالة المعالجة المسبقة.
Indexed sample sentence (first 10 elements): [15858, 2, 3250, 2495, 22224, 37864, 12367, 15513, 0, 0]
Indexed sample sentence (full length): 100

Original list of lists: [[1, 2, 3], [4, 5, 6], [7]]
Flattened list using concatenate: [1, 2, 3, 4, 5, 6, 7]


## Final Task

### Subtask:
Review the preprocessing steps applied using the new functions and confirm if further steps are needed.


## Summary:

### Q&A
The preprocessing steps applied using the new functions are confirmed to be successful and have prepared the text data for further analysis or machine learning tasks. No immediate further preprocessing steps are explicitly needed within the scope of this task, as the data has been transformed into a numerical, fixed-length format.

### Data Analysis Key Findings
*   The `LookupError` preventing Arabic text tokenization was addressed by defining custom preprocessing and tokenization functions that handle Arabic text normalization, diacritic removal, and stopword filtering.
*   New functions (`preprocess`, `tokenize`, `preprocess_input`, `concatenate`, `data_vocab`) were successfully defined and utilized, including silent NLTK downloads for `stopwords` and `punkt`.
*   A `TypeError` due to non-string values in the `text` column was resolved by using `df['text'].fillna('')` before applying the preprocessing steps.
*   A new `cleaned_text` column was successfully created by applying the `preprocess` function, which handles Tatweel, diacritics, and normalizes Arabic character forms.
*   A `tokenized_text` column was generated by applying the `tokenize` function, which splits text by whitespace and removes Arabic stopwords.
*   A comprehensive word frequency vocabulary was generated from the `tokenized_text`, resulting in a vocabulary size of 388,617 unique words.
*   Numerical representations (`word_to_id`, `id_to_word`) were created, and `indexed_text` columns were successfully generated, converting cleaned text into fixed-length (100) numerical sequences, ready for model input.
*   The `preprocess_input` function was successfully demonstrated, converting a sample Arabic sentence into its indexed numerical representation (e.g., `[15858, 2, 3250, 2495, 22224, 37864, 12367, 15513, 0, 0, ...]`).
*   The `concatenate` function was demonstrated to flatten a list of lists (e.g., `[[1, 2, 3], [4, 5, 6], [7]]` became `[1, 2, 3, 4, 5, 6, 7]`).

### Insights or Next Steps
*   The data is now thoroughly cleaned, tokenized, and converted into numerical sequences suitable for various machine learning models requiring fixed-length input.
*   Consider exploring advanced tokenization techniques or fine-tuning the stopword list if specific domain requirements or model performance issues arise.


# Task
Develop a sentiment analysis model for the provided Arabic text data by mapping the 'targe' column (0-4) to sentiment categories, extracting features from the 'indexed_text' column using TF-IDF, training a sentiment classifier, and developing a function to predict sentiment and explain the predictions by highlighting contributing words. Finally, review the model's performance and insights.

## Map Target Labels to Sentiment Categories

### Subtask:
Examine the unique values in the 'targe' column (0-4) and define a mapping to represent sentiment categories (e.g., negative, neutral, positive). This is crucial for training a sentiment classifier.


**Reasoning**:
First, I will inspect the unique values in the 'targe' column to understand the existing numerical labels before defining a mapping to sentiment categories. This is a crucial first step to ensure the mapping is correct.



In [ ]:
print("Unique values in 'targe' column:")
print(df['targe'].unique())

Unique values in 'targe' column:
[0 1 2 3 4]


**Reasoning**:
Now that the unique values in the 'targe' column have been identified, I will define a mapping for these numerical labels to descriptive sentiment categories, apply this mapping to create a new 'sentiment_category' column, and then display the DataFrame head and the value counts for the new column to verify the transformation.



In [ ]:
sentiment_mapping = {
    0: 'Negative',
    1: 'Somewhat Negative',
    2: 'Neutral',
    3: 'Somewhat Positive',
    4: 'Positive'
}

df['sentiment_category'] = df['targe'].map(sentiment_mapping)

print("DataFrame with new 'sentiment_category' column:")
print(df.head())

print("\nValue counts for 'sentiment_category':")
print(df['sentiment_category'].value_counts())

DataFrame with new 'sentiment_category' column:
                                                text  targe  \
0  بين أستوديوهات ورزازات وصحراء مرزوكة وآثار ولي...      0   
1  قررت النجمة الأمريكية أوبرا وينفري ألا يقتصر ع...      0   
2  أخبارنا المغربية الوزاني تصوير الشملالي ألهب ا...      0   
3  اخبارنا المغربية قال ابراهيم الراشدي محامي سعد...      0   
4  تزال صناعة الجلود في المغرب تتبع الطريقة التقل...      0   

                                        cleaned_text  \
0  بين استوديوهات ورزازات وصحراء مرزوكه واثار ولي...   
1  قررت النجمه الامريكيه اوبرا وينفري الا يقتصر ع...   
2  اخبارنا المغربيه الوزاني تصوير الشملالي الهب ا...   
3  اخبارنا المغربيه قال ابراهيم الراشدي محامي سعد...   
4  تزال صناعه الجلود في المغرب تتبع الطريقه التقل...   

                                      tokenized_text  \
0  [استوديوهات, ورزازات, وصحراء, مرزوكه, واثار, و...   
1  [قررت, النجمه, الامريكيه, اوبرا, وينفري, الا, ...   
2  [اخبارنا, المغربيه, الوزاني, تصوير, الشملالي, ...   
3  [اخبارنا,

## Extract Features for Sentiment Model

### Subtask:
Convert the 'indexed_text' column into numerical features using a method like TF-IDF (Term Frequency-Inverse Document Frequency). This will transform the text data into a format suitable for machine learning algorithms.


**Reasoning**:
To prepare the data for TF-IDF vectorization, I need to convert the numerical `indexed_text` column back into human-readable text strings. I will iterate through the indexed text lists and use the `id_to_word` dictionary to reconstruct the text, storing it in a new column called 'text_for_tfidf'.



In [ ]:
def indexed_to_text(indexed_list, id_to_word):
    # Filter out padding (0) and map IDs to words
    words = [id_to_word.get(idx, '') for idx in indexed_list if idx != 0]
    return ' '.join(words)

# Create the 'text_for_tfidf' column
df['text_for_tfidf'] = df['indexed_text'].apply(lambda x: indexed_to_text(x, id_to_word))

print("DataFrame with 'text_for_tfidf' column:")
print(df[['indexed_text', 'text_for_tfidf']].head())

DataFrame with 'text_for_tfidf' column:
                                        indexed_text  \
0  [18319, 5236, 88993, 42856, 6229, 15203, 278, ...   
1  [723, 6193, 558, 33825, 49564, 57, 6732, 4024,...   
2  [111, 8, 4914, 1696, 43506, 37844, 1450, 18, 9...   
3  [111, 8, 54, 1548, 19486, 6184, 808, 5800, 1, ...   
4  [3035, 1867, 62693, 6, 3655, 1738, 1633, 18140...   

                                      text_for_tfidf  
0  استوديوهات ورزازات وصحراء مرزوكه واثار وليلي ا...  
1  قررت النجمه الامريكيه اوبرا وينفري الا يقتصر ع...  
2  اخبارنا المغربيه الوزاني تصوير الشملالي الهب ا...  
3  اخبارنا المغربيه قال ابراهيم الراشدي محامي سعد...  
4  تزال صناعه الجلود المغرب تتبع الطريقه التقليدي...  


**Reasoning**:
Now that the `text_for_tfidf` column is created, I will import `TfidfVectorizer` from `sklearn.feature_extraction.text`, instantiate it, fit it to the 'text_for_tfidf' column, transform the text data into TF-IDF features, and print the shape of the resulting array.



In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Instantiate TfidfVectorizer
# You can adjust parameters like max_features, min_df, max_df, ngram_range as needed
tfidf_vectorizer = TfidfVectorizer(max_features=10000, ngram_range=(1, 2))

# Fit and transform the 'text_for_tfidf' column
X_tfidf = tfidf_vectorizer.fit_transform(df['text_for_tfidf'])

# Print the shape of the TF-IDF features
print(f"Shape of X_tfidf (number of documents, number of features): {X_tfidf.shape}")

Shape of X_tfidf (number of documents, number of features): (111728, 10000)


## Split Data and Train Sentiment Classifier

### Subtask:
Split the dataset into training and testing sets. Then, train a machine learning model (e.g., Logistic Regression) using the TF-IDF features and the defined sentiment categories as labels. This step will also include evaluating the model's performance.


**Reasoning**:
I will start by defining the feature matrix `X` using the `X_tfidf` data and the target vector `y` using the `sentiment_category` column from the DataFrame, then import necessary libraries and split the data into training and testing sets to prepare for model training.



In [ ]:
from sklearn.model_selection import train_test_split

# 1. Define X (features) and y (target labels)
X = X_tfidf
y = df['sentiment_category']

# 2. Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (89382, 10000)
Shape of X_test: (22346, 10000)
Shape of y_train: (89382,)
Shape of y_test: (22346,)


**Reasoning**:
Now that the data has been split into training and testing sets, I will import the necessary machine learning libraries, instantiate and train a Logistic Regression model, make predictions on the test set, and evaluate its performance using accuracy score and a classification report.



In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# 4. Instantiate a Logistic Regression model
# Using 'saga' solver for multiclass classification and large datasets, and increased max_iter for convergence.
log_reg_model = LogisticRegression(max_iter=1000, solver='saga', random_state=42)

# 5. Train the Logistic Regression model
print("Training Logistic Regression model...")
log_reg_model.fit(X_train, y_train)
print("Model training complete.")

# 6. Make predictions on the test set
y_pred = log_reg_model.predict(X_test)

# 7. Calculate and print the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print(f"\nAccuracy: {accuracy:.4f}")

# 8. Print the classification report
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Training Logistic Regression model...
Model training complete.

Accuracy: 0.9534

Classification Report:
                   precision    recall  f1-score   support

         Negative       0.96      0.94      0.95      2748
          Neutral       0.91      0.89      0.90      2847
         Positive       0.98      0.99      0.99      9304
Somewhat Negative       0.95      0.97      0.96      3346
Somewhat Positive       0.91      0.91      0.91      4101

         accuracy                           0.95     22346
        macro avg       0.94      0.94      0.94     22346
     weighted avg       0.95      0.95      0.95     22346



## Predict Sentiment and Generate Explanations

### Subtask:
Develop a function that takes new text input, preprocesses it, and uses the trained model to predict its sentiment. For the explanation, the function will identify and highlight words that strongly contribute to the predicted positive or negative sentiment, leveraging the model's internal weights or a simplified approach.


**Reasoning**:
I need to develop the `predict_and_explain_sentiment` function as described. This involves preprocessing the input text, transforming it with the TF-IDF vectorizer, predicting sentiment using the trained logistic regression model, and then identifying and highlighting words that contribute most to the prediction. I'll also retrieve feature names and coefficients to calculate word contributions.



In [ ]:
def predict_and_explain_sentiment(text, model, vectorizer, word_to_id, max_len, top_n=10):
    # 1. Preprocess and tokenize the input text
    cleaned_text = preprocess(text)
    tokens = tokenize(cleaned_text)

    # Join tokens back into a space-separated string for TF-IDF vectorization
    processed_text_for_tfidf = ' '.join(tokens)

    # 2. Transform the preprocessed text into TF-IDF features
    # tfidf_vectorizer expects an iterable, so pass in a list
    text_tfidf = vectorizer.transform([processed_text_for_tfidf])

    # 3. Predict the sentiment
    predicted_sentiment = model.predict(text_tfidf)[0]
    predicted_proba = model.predict_proba(text_tfidf)[0]

    # Find the index of the predicted sentiment in the model's classes
    sentiment_class_idx = list(model.classes_).index(predicted_sentiment)

    # 4. Get feature names from the TF-IDF vectorizer
    feature_names = vectorizer.get_feature_names_out()

    # 5. Extract coefficients corresponding to the predicted sentiment
    # Logistic Regression has one coef_ row per class in multiclass scenario
    coefs = model.coef_[sentiment_class_idx]

    # Map feature names to their coefficients for the predicted class
    feature_coef_map = {name: coef for name, coef in zip(feature_names, coefs)}

    # 6. Calculate word contributions for explanation
    contributions = []
    # Iterate through non-zero entries in the TF-IDF vector for the input text
    for col_idx in text_tfidf.nonzero()[1]:
        feature_name = feature_names[col_idx]
        tfidf_value = text_tfidf[0, col_idx]
        coefficient = feature_coef_map.get(feature_name, 0)
        contribution = tfidf_value * coefficient
        contributions.append((feature_name, contribution))

    # Sort contributions by magnitude
    contributions.sort(key=lambda x: abs(x[1]), reverse=True)

    # Separate positive and negative contributions
    positive_contributions = sorted([c for c in contributions if c[1] > 0], key=lambda x: x[1], reverse=True)[:top_n]
    negative_contributions = sorted([c for c in contributions if c[1] < 0], key=lambda x: x[1])[:top_n]

    print(f"Input Text: '{text}'")
    print(f"Predicted Sentiment: {predicted_sentiment}")
    print(f"Prediction Probability: {predicted_proba[sentiment_class_idx]:.4f}")

    print(f"\nTop {top_n} Positively Contributing Features for '{predicted_sentiment}':")
    if positive_contributions:
        for feature, score in positive_contributions:
            print(f"  - '{feature}': {score:.4f}")
    else:
        print("  No significant positive contributions found.")

    print(f"\nTop {top_n} Negatively Contributing Features for '{predicted_sentiment}':")
    if negative_contributions:
        for feature, score in negative_contributions:
            print(f"  - '{feature}': {score:.4f}")
    else:
        print("  No significant negative contributions found.")


# Test the function with a few sample Arabic sentences
sample_texts = [
    "هذا المنتج رائع جداً وأنا سعيد بشرائه.",
    "الخدمة كانت سيئة للغاية وغير مرضية.",
    "الجو اليوم معتدل ومناسب للخروج.",
    "لم يكن سيئا جداً لكنه لم يكن رائعاً أيضاً.",
    "أداء الفريق كان مذهلاً واستحق الفوز."
]

print("\n--- Testing Sentiment Prediction and Explanation ---")
for s_text in sample_texts:
    print("\n" + "="*50)
    predict_and_explain_sentiment(s_text, log_reg_model, tfidf_vectorizer, word_to_id, max_len)



--- Testing Sentiment Prediction and Explanation ---

Input Text: 'هذا المنتج رائع جداً وأنا سعيد بشرائه.'
Predicted Sentiment: Positive
Prediction Probability: 0.5392

Top 10 Positively Contributing Features for 'Positive':
  - 'رائع': 0.0199

Top 10 Negatively Contributing Features for 'Positive':
  - 'سعيد': -0.4064
  - 'المنتج': -0.3104
  - 'جدا': -0.1224
  - 'وانا': -0.0890

Input Text: 'الخدمة كانت سيئة للغاية وغير مرضية.'
Predicted Sentiment: Positive
Prediction Probability: 0.6198

Top 10 Positively Contributing Features for 'Positive':
  - 'للغايه': 0.0842

Top 10 Negatively Contributing Features for 'Positive':
  - 'كانت': -0.4590
  - 'الخدمه': -0.3168
  - 'وغير': -0.2638

Input Text: 'الجو اليوم معتدل ومناسب للخروج.'
Predicted Sentiment: Somewhat Positive
Prediction Probability: 0.3254

Top 10 Positively Contributing Features for 'Somewhat Positive':
  - 'اليوم': 1.1443

Top 10 Negatively Contributing Features for 'Somewhat Positive':
  No significant negative contributions

## Final Task

### Subtask:
Review the sentiment analysis results and explanations, discussing the model's performance and any insights gained.


## Summary:

### Data Analysis Key Findings

*   **Sentiment Category Mapping:** The numerical 'targe' column (0-4) was successfully mapped to five descriptive sentiment categories: 'Negative', 'Somewhat Negative', 'Neutral', 'Somewhat Positive', and 'Positive'. The dataset showed a prevalence of 'Positive' sentiment (46,522 instances), followed by 'Somewhat Positive' (20,505), 'Somewhat Negative' (16,728), 'Neutral' (14,235), and 'Negative' (13,738).
*   **TF-IDF Feature Extraction:** Text data was converted into 10,000 TF-IDF features using `TfidfVectorizer` with `max_features=10000` and `ngram_range=(1, 2)`, resulting in a feature matrix of shape (111,728, 10,000).
*   **High Model Accuracy:** A Logistic Regression model, trained on 80% of the data (89,382 samples), achieved an impressive accuracy of 95.35% on the remaining 20% test set (22,346 samples).
*   **Strong Performance Across Categories:** The classification report indicated high F1-scores for all sentiment categories, with 'Positive' sentiment achieving the highest at 0.99. 'Negative' scored 0.95, 'Somewhat Negative' 0.96, 'Somewhat Positive' 0.91, and 'Neutral' 0.90. Both macro and weighted average F1-scores were around 0.94-0.95.
*   **Explanation Functionality Developed:** A function was created to predict sentiment for new text and explain the prediction by highlighting top positively and negatively contributing words, based on TF-IDF values and model coefficients. While generally effective, one sample text "الخدمة كانت سيئة للغاية وغير مرضية." (The service was very bad and unsatisfactory.) was incorrectly predicted as 'Positive' during testing, despite its clear negative sentiment.

### Insights or Next Steps

*   The model demonstrates excellent overall performance for Arabic sentiment analysis. However, the identified misclassification of a clearly negative sentence as 'Positive' indicates a potential area for improvement, perhaps by examining the training data for similar patterns or exploring more advanced models for nuanced sentiment.
*   The developed explanation function is a valuable tool for understanding model predictions. For future development, integrating more sophisticated interpretability methods like LIME or SHAP could offer deeper insights into the word contributions and help diagnose model biases or errors more effectively.


In [ ]:
# Install necessary libraries
!pip install -qU langchain-google-genai langchain

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.5/66.5 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 111.2/111.2 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 496.3/496.3 kB 30.0 MB/s eta 0:00:00


You mentioned providing your Google API key. For security, it's recommended to store your API key in Colab's secrets manager. You can access it by clicking the "🔑" icon in the left sidebar and adding a new secret named `GOOGLE_API_KEY`.

In [ ]:
from langchain_google_genai import ChatGoogleGenerativeAI
from google.colab import userdata

# Retrieve the API key from Colab secrets
GOOGLE_API_KEY = userdata.get('GOOGLE_API_KEY')

# Initialize the Gemini model
llm = ChatGoogleGenerativeAI(model="gemini-2.5-flash", google_api_key=GOOGLE_API_KEY)

print("LangChain and Gemini model environment prepared.")

LangChain and Gemini model environment prepared.


In [ ]:
sample_text_for_gemini = "الخدمة كانت سيئة للغاية وغير مرضية." # The same text that Logistic Regression misclassified

# Create a prompt for Gemini to perform sentiment analysis and explain its reasoning
prompt = f"Perform sentiment analysis on the following Arabic text and explain your reasoning:\n\nText: '{sample_text_for_gemini}'\n\nSentiment: "

print(f"Sending prompt to Gemini:\n{prompt}")

# Get response from Gemini
gemini_response = llm.invoke(prompt)

print("\n--- Gemini's Sentiment Analysis ---")
print(gemini_response.content)

Sending prompt to Gemini:
Perform sentiment analysis on the following Arabic text and explain your reasoning:

Text: 'الخدمة كانت سيئة للغاية وغير مرضية.'

Sentiment: 

--- Gemini's Sentiment Analysis ---
**Sentiment:** Negative

**Reasoning:**

1.  **"سيئة" (sayyi'a):** This word directly translates to "bad" or "poor," which carries a strong negative connotation.
2.  **"للغاية" (lil-ghaya):** This phrase means "extremely" or "very," acting as an intensifier that amplifies the negative sentiment of "bad." So, "extremely bad."
3.  **"وغير مرضية" (wa-ghayr mardiya):** This translates to "and not satisfactory" or "and unsatisfactory." The word "غير" (ghayr) means "not" or "un-," negating the positive concept of "satisfactory" (مرضية), thereby adding another layer of negativity.

Combining these elements, the text clearly expresses a highly negative experience with the service.


In [ ]:
import ipywidgets as widgets
from IPython.display import display, clear_output

# Create a text area for input
text_input = widgets.Textarea(
    value='',
    placeholder='Enter Arabic text here...',
    description='Your Text:',
    disabled=False,
    layout=widgets.Layout(width='auto', height='100px')
)

# Create a button to trigger sentiment analysis
analyze_button = widgets.Button(
    description='Analyze Sentiment',
    disabled=False,
    button_style='info', # 'success', 'info', 'warning', 'danger' or ''
    tooltip='Click to analyze sentiment'
)

# Create output widgets for displaying results
output_lr = widgets.Output()
output_gemini = widgets.Output()

def on_analyze_button_clicked(b):
    with output_lr:
        clear_output()
        print("\n--- Logistic Regression Model Analysis ---")
        # Preprocess the input text for LR model
        cleaned_text_lr = preprocess(text_input.value)
        tokens_lr = tokenize(cleaned_text_lr)
        processed_text_for_tfidf_lr = ' '.join(tokens_lr)

        if processed_text_for_tfidf_lr:
            text_tfidf_lr = tfidf_vectorizer.transform([processed_text_for_tfidf_lr])
            predicted_sentiment_lr = log_reg_model.predict(text_tfidf_lr)[0]
            predicted_proba_lr = log_reg_model.predict_proba(text_tfidf_lr)[0]
            sentiment_class_idx_lr = list(log_reg_model.classes_).index(predicted_sentiment_lr)

            print(f"Predicted Sentiment: {predicted_sentiment_lr}")
            print(f"Prediction Probability: {predicted_proba_lr[sentiment_class_idx_lr]:.4f}")

            # Explanation for LR model (re-using the function from before)
            predict_and_explain_sentiment(text_input.value, log_reg_model, tfidf_vectorizer, word_to_id, max_len)
        else:
            print("No meaningful words found after preprocessing for Logistic Regression.")

    with output_gemini:
        clear_output()
        print("\n--- Gemini LLM Analysis ---")
        prompt_gemini = f"Perform sentiment analysis on the following Arabic text and explain your reasoning:\n\nText: '{text_input.value}'\n\nSentiment: "
        try:
            gemini_response = llm.invoke(prompt_gemini)
            print(gemini_response.content)
        except Exception as e:
            print(f"Error calling Gemini LLM: {e}")


# Attach the event handler to the button
analyze_button.on_click(on_analyze_button_clicked)

# Display the widgets
print("Enter text and click 'Analyze Sentiment' to see predictions:")
display(text_input, analyze_button, output_lr, output_gemini)


Enter text and click 'Analyze Sentiment' to see predictions:


Textarea(value='', description='Your Text:', layout=Layout(height='100px', width='auto'), placeholder='Enter A…

Button(button_style='info', description='Analyze Sentiment', style=ButtonStyle(), tooltip='Click to analyze se…

Output()

Output()

# Task
Install the `transformers` library and import `AutoTokenizer`, `AutoModelForSequenceClassification`, and `TrainingArguments` to prepare for fine-tuning a BERT-based model for sentiment analysis.

## Install and Import BERT-related Libraries

### Subtask:
Install the `transformers` library from Hugging Face and import necessary components like `AutoTokenizer`, `AutoModelForSequenceClassification`, and `TrainingArguments`.


**Reasoning**:
To begin the subtask, I need to install the `transformers` library which contains the necessary components for BERT-related tasks, and then import `AutoTokenizer`, `AutoModelForSequenceClassification`, and `TrainingArguments`.



In [ ]:
import warnings
warnings.filterwarnings('ignore')

# Install the transformers library
!pip install -q transformers

# Import necessary components
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments

print("Transformers library installed and components imported.")

Transformers library installed and components imported.


## Load Pre-trained Arabic BERT Tokenizer and Model

### Subtask:
Load a suitable pre-trained Arabic BERT tokenizer and a corresponding model with a sequence classification head.


**Reasoning**:
First, I will determine the number of unique sentiment categories from the 'sentiment_category' column in the DataFrame to set the 'num_labels' for the BERT model, then load a pre-trained Arabic BERT tokenizer and model using the specified model name and the determined number of labels, and finally print the loaded tokenizer and model.



In [ ]:
model_name = "UBC-NLP/MARBERTv2"

# 1. Determine the number of unique sentiment categories
num_labels = df['sentiment_category'].nunique()

# 2. Load the tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name)

# 3. Load the pre-trained model with a sequence classification head
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)

print(f"Number of unique sentiment labels: {num_labels}")
print("Tokenizer loaded successfully:", tokenizer.__class__.__name__)
print("Model loaded successfully:", model.__class__.__name__)

config.json:   0%|          | 0.00/757 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/439 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/654M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/654M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertForSequenceClassification LOAD REPORT from: UBC-NLP/MARBERTv2
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
cls.predictions.transform.dense.bias       | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
cls.predictions.bias                       | UNEXPECTED | 
cls.predictions.decoder.weight             | UNEXPECTED | 
cls.predictions.decoder.bias               | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
bert.embeddings.position_ids               | UNEXPECTED | 
classifier.bias                            | MISSING    | 
classifier.weight                          | MISSING    | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect ide

Number of unique sentiment labels: 5
Tokenizer loaded successfully: BertTokenizer
Model loaded successfully: BertForSequenceClassification


**Reasoning**:
The previous step successfully loaded the tokenizer and model, with expected warnings about the newly initialized classification head. The next logical step is to prepare the data for fine-tuning the BERT model. This involves mapping the sentiment categories to numerical labels and then tokenizing the text data using the loaded `tokenizer`.



In [ ]:
from sklearn.preprocessing import LabelEncoder

# 1. Map sentiment categories to numerical labels
label_encoder = LabelEncoder()
df['labels'] = label_encoder.fit_transform(df['sentiment_category'])

# Display the mapping and the new 'labels' column
sentiment_id_map = dict(zip(label_encoder.classes_, label_encoder.transform(label_encoder.classes_)))
print(f"Sentiment to ID mapping: {sentiment_id_map}")
print(f"ID to Sentiment mapping: {dict(zip(label_encoder.transform(label_encoder.classes_), label_encoder.classes_))}")

print("\nDataFrame with 'labels' column:")
print(df[['sentiment_category', 'labels']].head())

# 2. Tokenize the 'cleaned_text' column
def tokenize_function(examples):
    return tokenizer(examples['cleaned_text'], truncation=True, padding='max_length', max_length=128)

# To apply tokenization, we'll convert the DataFrame into a Hugging Face Dataset format temporarily
# This allows for efficient batch processing of tokenization
from datasets import Dataset

# Create a Hugging Face Dataset from the relevant columns
hf_dataset = Dataset.from_pandas(df[['cleaned_text', 'labels']])

# Apply the tokenization function
tokenized_datasets = hf_dataset.map(tokenize_function, batched=True)

print("\nTokenized dataset features:")
print(tokenized_datasets[0].keys())
print("First tokenized example (input_ids):")
print(tokenized_datasets[0]['input_ids'][:10])
print("First tokenized example (attention_mask):")
print(tokenized_datasets[0]['attention_mask'][:10])
print("First tokenized example (labels):")
print(tokenized_datasets[0]['labels'])


Sentiment to ID mapping: {'Negative': np.int64(0), 'Neutral': np.int64(1), 'Positive': np.int64(2), 'Somewhat Negative': np.int64(3), 'Somewhat Positive': np.int64(4)}
ID to Sentiment mapping: {np.int64(0): 'Negative', np.int64(1): 'Neutral', np.int64(2): 'Positive', np.int64(3): 'Somewhat Negative', np.int64(4): 'Somewhat Positive'}

DataFrame with 'labels' column:
  sentiment_category  labels
0           Negative       0
1           Negative       0
2           Negative       0
3           Negative       0
4           Negative       0


Map:   0%|          | 0/111728 [00:00<?, ? examples/s]


Tokenized dataset features:
dict_keys(['cleaned_text', 'labels', 'input_ids', 'token_type_ids', 'attention_mask'])
First tokenized example (input_ids):
[2, 2199, 54125, 21826, 39620, 3358, 1949, 11271, 2394, 2510]
First tokenized example (attention_mask):
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
First tokenized example (labels):
0


**Reasoning**:
Now that the data is tokenized and labeled, I will split the `tokenized_datasets` into training and testing sets. This is a crucial step before defining the training arguments and fine-tuning the BERT model to ensure proper evaluation of the model's performance.



In [ ]:
from sklearn.model_selection import train_test_split

# Convert tokenized_datasets to pandas DataFrame for easier splitting with train_test_split
# This is a temporary conversion as Hugging Face Trainer expects a Dataset object later
df_tokenized = tokenized_datasets.to_pandas()

# Split the DataFrame into training and testing sets
# We'll need to create Hugging Face Dataset objects from these pandas DataFrames later
train_df, test_df = train_test_split(
    df_tokenized, test_size=0.2, random_state=42, stratify=df_tokenized['labels']
)

# Convert back to Hugging Face Dataset objects for the Trainer
train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df)

print(f"Training dataset size: {len(train_dataset)}")
print(f"Test dataset size: {len(test_dataset)}")

# Remove the '__index_level_0__' column that gets added during to_pandas() conversion
# This column can cause issues with the Hugging Face Trainer
train_dataset = train_dataset.remove_columns(['__index_level_0__'])
test_dataset = test_dataset.remove_columns(['__index_level_0__'])

print("Columns in training dataset after cleanup:", train_dataset.column_names)
print("Columns in test dataset after cleanup:", test_dataset.column_names)

Training dataset size: 89382
Test dataset size: 22346
Columns in training dataset after cleanup: ['cleaned_text', 'labels', 'input_ids', 'token_type_ids', 'attention_mask']
Columns in test dataset after cleanup: ['cleaned_text', 'labels', 'input_ids', 'token_type_ids', 'attention_mask']


**Reasoning**:
With the data now split into training and testing sets, I will define the training arguments for the BERT model using `TrainingArguments` and then initialize the `Trainer` object. This sets up the environment for fine-tuning the model.



### Comparison with Logistic Regression Model

**Logistic Regression Model Performance:**
*   **Accuracy:** 0.9534
*   **Classification Report (F1-scores):**
    *   Negative: 0.95
    *   Neutral: 0.90
    *   Positive: 0.99
    *   Somewhat Negative: 0.96
    *   Somewhat Positive: 0.91
    *   Macro Avg F1: 0.94
    *   Weighted Avg F1: 0.95

**BERT Model Performance (from `eval_results`):**

(The BERT model's evaluation results will be displayed above this section after execution.)

**Discussion:**

After the BERT model evaluation is run, we will compare the `eval_results` (specifically `eval_accuracy` and potentially other metrics if calculated by `compute_metrics`) with the Logistic Regression metrics listed above. This comparison will highlight whether fine-tuning a BERT-based model provided a significant improvement in sentiment classification accuracy and other performance indicators for this Arabic text dataset.

**Reasoning**:
Now that the BERT model training is complete, I will call the `trainer.evaluate()` method to assess its performance on the test set and retrieve relevant metrics for comparison.



In [ ]:
pip install -q evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 4.1 MB/s eta 0:00:00


**Reasoning**:
Since the `evaluate` library has now been installed, I will re-run the previous code block to correctly import `evaluate.load` and initialize the Hugging Face `Trainer` object, allowing for the BERT model fine-tuning setup to proceed.



In [ ]:
import numpy as np
from transformers import Trainer
import evaluate # Correct import for metrics

# 1. Define TrainingArguments
# These parameters control the training process
training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,  # Number of training epochs
    per_device_train_batch_size=16, # Batch size for training
    per_device_eval_batch_size=16,  # Batch size for evaluation
    warmup_steps=500,                # Number of warmup steps for learning rate scheduler
    weight_decay=0.01,               # Strength of weight decay
    logging_dir='./logs',            # Directory for storing logs
    logging_steps=1000,              # Log every X updates steps
    eval_strategy='epoch',     # Evaluate the model at the end of each epoch (changed from evaluation_strategy)
    save_strategy='epoch',           # Save the model at the end of each epoch
    load_best_model_at_end=True,     # Load the best model at the end of training
    metric_for_best_model='accuracy',# Metric to use for best model selection
    report_to='none'                 # Disable reporting to services like Weights & Biases
)

# 2. Define a function to compute metrics for evaluation
# This function will be passed to the Trainer
def compute_metrics(eval_pred):
    metric = evaluate.load("accuracy") # Use evaluate.load instead of datasets.load_metric
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

# 3. Initialize the Trainer
# The Trainer takes the model, training arguments, train/eval datasets, and compute_metrics function
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    compute_metrics=compute_metrics,
)

print("Evaluating fine-tuned BERT model...")
eval_results = trainer.evaluate()
print("BERT model evaluation complete.")

print("\nBERT Model Evaluation Results:")
for key, value in eval_results.items():
    print(f"{key}: {value:.4f}")

`logging_dir` is deprecated and will be removed in v5.2. Please set `TENSORBOARD_LOGGING_DIR` instead.


Evaluating fine-tuned BERT model...


BERT model evaluation complete.

BERT Model Evaluation Results:
eval_loss: 1.6240
eval_model_preparation_time: 0.0132
eval_accuracy: 0.1243
eval_runtime: 164.5073
eval_samples_per_second: 135.8360
eval_steps_per_second: 8.4920


**Reasoning**:
The previous evaluation of the BERT model showed extremely low accuracy, indicating that the model was not yet trained. To properly evaluate the *fine-tuned* BERT model as per the subtask, I must first execute the training process. The `trainer.train()` method is located in the next logical code block and needs to be run before evaluation.



In [ ]:
print("Starting model training...")
trainer.train()
print("Model training complete.")

Starting model training...


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

KeyboardInterrupt: 